specs/consensus/overview.md

```diff
@@ -437,18 +437,134 @@ height of consensus `h_p`.
 
 ### Validation
 
-The external function `valid(v)` is invoked by a process when it receives a
-`⟨PROPOSAL, h, r, v, *⟩` from the proposer of round `r` of height `h`.
-It should return whether the `v` is a valid value according to the semantics of 
-the "client" of the consensus protocol, i.e., the application that uses
-consensus to agree on proposed values.
-
-> TODO: relevant observation:
-> - Validation typically depends on `h` as well, in particular on the
->   application state at blockchain height `h`. It should not ordinarily
->   depend on `r`, since the application state should not change over rounds.
-> - Determinism: is `valid(v)` a function?
-> - Needed because of validity property of consensus, defined in the paper
+The external function `valid` is used to guard any value-specific action in 
+the algorithm. The purpose is to restrict the domain of values that consensus 
+may decide. There is the assumption that if `v` is a value returned
+by the  `getValue()` function of a correct process in the current height, then
+`valid(v)` evaluates to `true`. Under this assumption, as long as the proposer
+is correct, `valid(v)` does not serve any purpose. It only becomes relevant if
+there is a faulty proposer; it limits "how bad" proposed values can be.
+
+In the arXiv paper, the pseudo code uses `valid(v)`, that is, a function that
+maps a value to a boolean. Observe that this should be understood in terms
+of a mathematical (pure) function:
+1. the function must only depend on `v` but not on other data 
+(e.g., an application state at a given point in time, the current local time 
+or temperature).
+2. If invoked several times for the same input, the function always returns 
+the same boolean value (determinism).
+3. If invoked on different processes for the same input, the function always 
+returns the same boolean value. 
+
+The correctness of the consensus algorithm depends heavily on these points. 
+Most obviously for termination, we require that all correct processes find 
+the value by a correct proposer valid under all circumstances, because we
+need them to all prevote for the value. A deviation in `valid` would 
+result in some processes
+prevoting nil, and consequently, there might never be enough votes for a 
+value to decide.
+
+#### `valid(v)` in implementations
+
+In implementations, the validity check typically is not pure:
+as already mentioned in the arXiv paper "In the context of blockchain systems, 
+for example, a value is not valid if it does not
+contain an appropriate hash of the last value (block) added to the blockchain." That is, 
+rather than `valid(v)` the implementation looks something like `valid(v, chain, height)`, 
```

Comment:
Regarding `valid(v, chain, height)`, I agree with the discussion but I would use the nomenclature of the pseudo-code.

Instead of `chain`, we have the `decision_p` array for process `p`, and instead of `height` we use `h_p`.

The height here can actually be omitted, as by construction if `h_p = H`, then `decision_p` is filled from `0` to `H-1`. In other words, `len(decision_p) = h_p` is an invariant.